File: trunk/examples/app-fermi/2.0/fermi.py
# ----------------------------------------------------------------------
#  FERMI STATISTICS
#
#  This simple example shows how you can use the Rappture toolkit
#  to handle I/O for a simple simulator--in this case, one that
#  computes Fermi-Dirac statistics.
#
# ======================================================================
#  AUTHOR:  Michael McLennan, Purdue University
#  Copyright (c) 2004-2012  HUBzero Foundation, LLC
#
#  See the file "license.terms" for information on usage and
#  redistribution of this file, and for a DISCLAIMER OF ALL WARRANTIES.
# ======================================================================

from math import *
import sys
import Rappture

import fermi_io
from fermi_io import *

# do this right at the start, to handle command line args
Rappture.interface(sys.argv, fermi_io)

kT = 8.61734e-5 * T
Emin = Ef - 10*kT
Emax = Ef + 10*kT

E = Emin; dE = 0.005*(Emax-Emin)

Rappture.driver.put("output.curve(f12).about.label", "Fermi-Dirac Factor")
Rappture.driver.put("output.curve(f12).xaxis.label", "Energy")
Rappture.driver.put("output.curve(f12).xaxis.units", "eV")
path = "output.curve(f12).component.xy"

while E < Emax:
    f = 1.0/(1.0 + exp((E - Ef)/kT))
    # result.append( [E,f] )
    value = "%s  %s\n" % (f,E)
    Rappture.driver.put(path,value,append=1)
    E += dE

File: trunk/examples/app-fermi/wrapper/perl/fermi.pl
# ----------------------------------------------------------------------
#  EXAMPLE: Fermi-Dirac function in Perl.
#         
#  This simple example shows how to use Rappture within a simulator
#  written in Perl.
# ======================================================================
#  AUTHOR:  Nicholas J. Kisseberth, Purdue University
#  Copyright (c) 2004-2012  HUBzero Foundation, LLC
#
#  See the file "license.terms" for information on usage and
#  redistribution of this file, and for a DISCLAIMER OF ALL WARRANTIES.
# ======================================================================

use Rappture;

# open the XML file containing the run parameters

$driver = Rappture::RpLibrary->new($ARGV[0]);

$Tstr = $driver->get("input.(temperature).current");
$T = Rappture::RpUnits::convert($Tstr, "K", "off");

$Efstr = $driver->get("input.(Ef).current");
$Ef = Rappture::RpUnits::convert($Efstr, "eV", "off");

open FILE, ">indeck" or die $!;
print FILE "$Ef\n";
print FILE "$T\n";
close FILE;

# $result = `octave --silent fermi.m < indeck 2>&1`;
$result = `octave --silent fermi.m < indeck`;

$driver->put("output.log",$result,0);

if (length($result) > 0) {
    open FILE, "<out.dat" or die $!;
    my @lines = <FILE>;
    close FILE;


    # Label the output graph with a title, x-axis label,
    # y-axis label, y-axis units.

    $driver->put("output.curve(f12).about.label","Fermi-Dirac Factor",0);
    $driver->put("output.curve(f12).xaxis.label","Fermi-Dirac Factor",0);
    $driver->put("output.curve(f12).yaxis.label","Energy",0);
    $driver->put("output.curve(f12).yaxis.units","eV",0);

    # skip over the first 4 header lines
    split(@lines,0,4);
    # search each line for 2 floating point numbers
    foreach $line (@lines) {
        if ($line =~ /(?<f>[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?)\s+(?<E>[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?)/) {
            $driver->put("output.curve(f12).component.xy", "$+{f} $+{E}\n", 1);
        }
    }

}

unlink("indeck");
unlink("out.dat");

$driver->result();
